### ISSUE TITLE
running_loss的计算以及清空方式或存在重大Bug?

### ISSUE THREAD

**Issue by xinfeng1i** [Morizeyao/GPT2-Chinese#70]
https://github.com/Morizeyao/GPT2-Chinese/blob/130f04b5b6ad40f20ca28c519483eb4ca3e0b13b/train.py#L212-L229

上面是原来 train.py 代码中计算 running_loss 的方式，其中正常的处理逻辑应该如下：

1. 计算每个 batch 的 loss，并且更新梯度（但是不更新模型参数）;
2. 将每个 batch 的 loss 累加到 running_loss;
3. 如果达到 gradient_accumulation  的要求，则对调用优化器，对前面类累加的梯度进行模型参数更新;
4. 如果达到了要打印日志的步骤，则打印必要的日志 。

按照上述逻辑，目前的代码存在问题如下：

- 问题1.  第 214 行的 runing_loss 的更新，应该在第 213 行之前
- 问题2.  第 229 行 running_loss  的清空方式很奇怪：a) 第一个为什么要清空 running_loss ? b) 即便要清空 running_loss，也不应该是在打印日志循环中清空，打不打印日志本身不应该影响代码的训练逻辑，只是为了方便观察而已；而现在的处理方式，打不打印日志明显影响到了训练的结果 （running_loss是否清空）
- 问题3.  平均 loss 的计算方式，通常来说在训练中我们打印的 loss 应该是_训练到目前为止，平均每个 Example或者Batch的损失_，但是现在的计算方式似乎仅告诉_当前这个 Batch 的 Loss_，这是不对的。问题3+问题2或许会间接导致了 Issue #19 的出现。 

关于问题1，问题2的解决可以参考 https://github.com/huggingface/pytorch-transformers/blob/7c0f2d0a6a8937063bb310fceb56ac57ce53811b/examples/run_lm_finetuning.py#L202-L211

**Comment by lioyou** [comment]
1. Issue#19，我自己重构的版本，依旧会存在，当然并不会出现很大的波动，所以我觉得跟数据本身也有关系。另外，这些问题，我前面也有提过#60，所以自己重构就好了。
2. 之前那个total_step，在训练之前，batch_size与seq_len(n_ctx)是已知的，所以完全可以这样做：
> step_len = batch_size * stride + seq_len - stride
> full_len // step_len 这样子，最多丢失step_len长度的数据，当然也可以不丢失
> 最后训练时，以行去读取，每一行就是一个step
所以可以按一个batch_total_len=step_len来分割数据，另外stride，我觉得应该取seq_len(n_ctx)-1

**Comment by lioyou** [comment]
> 问题3. 平均 loss 的计算方式，通常来说在训练中我们打印的 loss 应该是_训练到目前为止，平均每个 Example或者Batch的损失_，但是现在的计算方式似乎仅告诉_当前这个 Batch 的 Loss_，这是不对的。

我觉得你是对的，就是数据原因及loss平均问题，因为计算只针对单个batch的loss，而非全部batch的loss平均值，那么当输入的数据存在较大的差异时，此时loss会波动，而打印的是单个batch的loss，所以某个batch数据差异非常大时，就会出现loss异常升高。这跟分片与否关系不大，与问题2也无关，因为此时不用梯度累积问题依旧在，而是跟数据的差异性有关，体现在平均loss与单个loss的计算问题上。
另外为什么只关注换片时的loss，是因为打印的时候，换片位置明显，其实其它的位置也会有波动，我观察了我的训练情况，的确是这样的。
谢谢，愉快的玩耍去了~

**Comment by lioyou** [comment]
以下是平均loss的结果（之前我做过同样的测试，但是有波动，唯一的改变就是平均loss的计算）
![image](https://user-images.githubusercontent.com/17487200/65566093-70a43e80-df84-11e9-94f5-3a52d090c1a4.png)
所以的确就是数据本身及平均loss的计算问题，另外我的模型并没有使用梯度累积，不存在问题1及2，所以loss异常升高就是平均loss及数据本身的差异。
可以关闭 #19
谢谢！

**Comment by Morizeyao** [comment]
我看一下

**Comment by Morizeyao** [comment]
问题3你说的是对的，确实会导致loss波动的问题。
对于问题12，现在代码对于梯度积累确实缺乏测试，我在实际使用中也没有使用到这个功能，我需要仔细斟酌一下。打印loss的话，目前没有设置不print的功能。
